src/Bicep.Core/Semantics/Namespaces/DefaultNamespaceProvider.cs

```diff
@@ -14,60 +16,60 @@ namespace Bicep.Core.Semantics.Namespaces;
-                    // performance hit as a result of recreating the provider for each call to TryGetNamespace.
-                    // Tracked by - https://msazure.visualstudio.com/One/_sprints/taskboard/Azure-ARM-Deployments/One/Custom/Azure-ARM/Gallium/Jul-2023?workitem=24563440
-                    var loader = azResourceTypeLoaderFactory.GetResourceTypeLoader(version, features);
-                    if (loader is null)
-                    {
-                        return null;
-                    }
-                    var overriddenProviderVersion = builtInAzResourceTypeLoaderVersion;
-                    if (features.DynamicTypeLoadingEnabled)
-                    {
-                        overriddenProviderVersion = version ?? overriddenProviderVersion;
-                    }
-                    provider = new AzResourceTypeProvider(loader, overriddenProviderVersion);
-                }
-                return AzNamespaceType.Create(alias, scope, provider, sourceFileKind);
-            },
-            [K8sNamespaceType.BuiltInName] = (alias, scope, features, sourceFileKind, version) => K8sNamespaceType.Create(alias),
-            [MicrosoftGraphNamespaceType.BuiltInName] = (alias, scope, features, sourceFileKind, version) => MicrosoftGraphNamespaceType.Create(alias),
+            [AzNamespaceType.BuiltInName] = TryCreateAzNamespace,
+            [SystemNamespaceType.BuiltInName] = (providerDescriptor, resourceScope, features, sourceFileKind) => SystemNamespaceType.Create(providerDescriptor.Alias, features, sourceFileKind),
+            [K8sNamespaceType.BuiltInName] = (providerDescriptor, resourceScope, features, sourceFileKind) => K8sNamespaceType.Create(providerDescriptor.Alias),
+            [MicrosoftGraphNamespaceType.BuiltInName] = (providerDescriptor, resourceScope, features, sourceFileKind) => MicrosoftGraphNamespaceType.Create(providerDescriptor.Alias),
         }.ToImmutableDictionary();
     }
 
+    private NamespaceType? TryCreateAzNamespace(ResourceTypesProviderDescriptor providerDescriptor, ResourceScope scope, IFeatureProvider features, BicepSourceFileKind sourceFileKind)
+    {
+        if (!features.DynamicTypeLoadingEnabled)
+        {
+            return AzNamespaceType.Create(
+                providerDescriptor.Alias,
+                scope,
+                resourceTypeLoaderFactory.GetBuiltInAzResourceTypesProvider(),
+                sourceFileKind);
+        }
+
+        if (!resourceTypeLoaderFactory.GetResourceTypeProvider(providerDescriptor, features).IsSuccess(out var dynamicallyLoadedProvider, out var errorBuilder))
+        {
+            Trace.WriteLine($"Failed to load types from {providerDescriptor.Path}: {errorBuilder(DiagnosticBuilder.ForPosition(providerDescriptor.Span))}");
+            return null;
+        }
+        return AzNamespaceType.Create(
+              providerDescriptor.Alias,
+              scope,
+              dynamicallyLoadedProvider!,
```

Comment:
I generally recommend avoiding `!` in product code as it can cover up problems. Is there a reason you needed this here?

Reply:
its not needed, I removed it